### PR TITLE
Implement audio stream playback parameters.

### PR DESCRIPTION
**Pull request by reduz**
Implements a way for audio stream playback to be configured via parameters directly in the edited AudioStreamPlayer[2D/3D].

Currently, configuring the playback stream is not possible (or is sometimes hacky as the user has to obtain the currently played stream, which is not always immediately available).

All you can do is configure the AudioStream itself, which affects all instances in the game. With this PR, configuring playback parameters per each stream player is possible, giving users far more flexibility and freedom to configure playback or animate parameters.

This PR only implements this new feature to control looping in stream playback instances (a commonly requested feature, which was lost in the transition from Godot 2 to Godot 3). But the idea is that it can do a lot more:

* If effects are bundled to the stream, control per playback instance parameters such as cutoff or resoance, or any other exposed effect parameter per playback instance.
* For the upcoming interactive music PR (#64488), this exposes an easy way to change the active clip, which was not possible before (so this PR is a prerequisite for that one).
* For the upcoming parametrizable audio support (https://github.com/godotengine/godot-proposals/issues/3394) this allows editing and animating audio graph parameters.

Screenshot of how it looks:

image

Supersedes #65797, implements https://github.com/godotengine/godot-proposals/issues/3120.

### CODE REVIEW

File: modules/vorbis/audio_stream_ogg_vorbis.h
Comment (AThousandShips) [inline modules/vorbis/audio_stream_ogg_vorbis.h:159]:
```suggestion
	virtual void get_parameter_list(List<Parameter> *r_parameters) override;
```

File: modules/minimp3/audio_stream_mp3.h
Comment (AThousandShips) [inline modules/minimp3/audio_stream_mp3.h:133]:
```suggestion
	virtual void get_parameter_list(List<Parameter> *r_parameters) override;
```

File: modules/minimp3/audio_stream_mp3.h
Comment (AThousandShips) [inline modules/minimp3/audio_stream_mp3.h:77]:
```suggestion
	virtual void set_parameter(const StringName &p_name, const Variant &p_value) override;
	virtual Variant get_parameter(const StringName &p_name) const override;
```

File: modules/vorbis/audio_stream_ogg_vorbis.h
Comment (AThousandShips) [inline modules/vorbis/audio_stream_ogg_vorbis.h:100]:
```suggestion
	virtual void set_parameter(const StringName &p_name, const Variant &p_value) override;
	virtual Variant get_parameter(const StringName &p_name) const override;
```

**Comment by KoBeWi** [comment]
Supersedes #65797 I guess.

File: doc/classes/AudioStream.xml
Comment (AThousandShips) [inline doc/classes/AudioStream.xml:34]:
```suggestion
				Return the controllable parameters of this stream. This array must be interleaved as a dictionary with a property info description (see [method Object.get_property_list]) and the default value for it.
```
Or similar, as it should return an array and not a count

**Comment by reduz** [comment]
@AThousandShips Thanks!! Merged and squashed.

**Comment by reduz** [comment]
Did some changes, looping is now something you can check to override, to make it less confusing.

I think this is ready for review and merge.

**Comment by aaronfranke** [comment]
I actually implemented per-player audio looping in The Mirror's fork of Godot a few months ago so that audio players can be looped by users, but I didn't know if this was desired to have upstream. So from that perspective, it's great to have this in the engine, less stuff for us to maintain :P

From the GLTF perspective, the latest iteration of the audio standard has looping as a part of the audio source (what Godot would call the audio stream), so this PR isn't useful for GLTF audio, but it doesn't hurt it either.

**Comment by reduz** [comment]
@aaronfranke The problem we have in the audio side is that parameters exposed at stream level are very tricky because:
1. Streams many times are imported, so you can't edit them directly.
2. Even if you can edit them directly, they are shared across all players.

So, this makes editing or interacting with local instances very hard. This PR is created to make this easy and allow exposing parameters that can be edited locally. Additionally, as they appear at node level (similar to how animationtree works), you can animate them and edit them more freely.

File: servers/audio/audio_stream.h
Comment (KoBeWi) [inline servers/audio/audio_stream.h:134]: Can this use TypedArray?

File: doc/classes/AudioStream.xml
Comment (KoBeWi) [inline doc/classes/AudioStream.xml:34]: I think I don't fully get this description. Is the expected format `[ Dictionary, default, Dictionary, default]`? Why the default can't be part of the Dictionary? PropertyInfo would just ignore it.

File: doc/classes/AudioStream.xml
Comment (KoBeWi) [inline doc/classes/AudioStream.xml:74]:
```suggestion
				Signal to be emitted to notify when the parameter list has changed.
```

**Comment by KoBeWi** [comment]
When you set a parameter in AudioStreamPlayer and then uncheck it, it will be stored in the scene as null.
image

File: scene/2d/audio_stream_player_2d.cpp
Comment (AThousandShips) [inline scene/2d/audio_stream_player_2d.cpp:467]: I think needs to detect if the variant is null, to handle the above issue

File: modules/minimp3/audio_stream_mp3.cpp
Comment (KoBeWi) [inline modules/minimp3/audio_stream_mp3.cpp:184]: Why is this set if it's the default value?

File: modules/minimp3/audio_stream_mp3.cpp
Comment (KoBeWi) [inline modules/minimp3/audio_stream_mp3.cpp:257]: And here. `Variant()` is the default value for the constructor.

File: modules/minimp3/audio_stream_mp3.h
Comment (KoBeWi) [inline modules/minimp3/audio_stream_mp3.h:51]: This could be an enum. `looping_override = false` and `looping = true` is invalid state.
Though it's fine like this too.

File: modules/vorbis/audio_stream_ogg_vorbis.cpp
Comment (KoBeWi) [inline modules/vorbis/audio_stream_ogg_vorbis.cpp:129]:
```suggestion
			if (beat_loop && is_not_empty) {
```
Though renaming the variable to `use_loop` like in MP3 would be more clear.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (KoBeWi) [inline modules/minimp3/audio_stream_mp3.cpp:160]:
```suggestion
	if (looping_override && p_name == SNAME("looping")) {
```
Cheaper check first.
Same in OGG.

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:39]: Can this be defined elsewhere, to avoid repeating it for every player type?
Maybe it could be `constexpr String` (or `char *` if String can't be used) in AudioStream?

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:228]:
```suggestion
	if (stream.is_null()) {
		return;
	}
```
Can return early.

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:232]:
```suggestion
			const PropertyInfo &pi = K.property;
```

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:466]:
```suggestion
	HashMap<StringName, ParameterData>::Iterator I = playback_parameters.find(p_name);
	if (I) {
		ParameterData &pd = I->value;
```
Avoids double lookup.

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:469]:
```suggestion
		for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
			playback->set_parameter(pd.path, pd.value);
```

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:478]: Same thing with iterator here.

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [inline scene/2d/audio_stream_player_2d.cpp:486]: This can technically also be an early return, though not sure if it fits this method.

**Comment by KoBeWi** [comment]
I reviewed the PR from the editor side and it looks good (aside from the null bug mentioned above).

Though I don't like how this adds even more code duplication to AudioStreamPlayer classes. I think they can be refactored using some internal class that handles common code.
That said, remember that my AudioStreamPlayer2D comments apply to other player too.

**Comment by Mickeon** [comment]
> Though I don't like how this adds even more code duplication to AudioStreamPlayer classes. I think they can be refactored using some internal class that handles common code.

Someday. Someday... They should eventually share the same interface. It would save a lot when it comes to code maintenance.
If https://github.com/godotengine/godot-proposals/issues/6416 were implemented even internally these 3 classes would probably be the ones that benefit it the most

**Comment by KoBeWi** [comment]
It doesn't need traits.

Maybe I'll get onto this once this PR is merged.

File: doc/classes/AudioStream.xml
Comment (reduz) [reply doc/classes/AudioStream.xml:34]: Thats a good point, will change to that.

**Comment by reduz** [comment]
@KoBeWi I agree, I think it's time we eventually create a shared internal class for all of them, though I guess not on the scope for this PR.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (reduz) [reply modules/minimp3/audio_stream_mp3.cpp:257]: That is okay because It's a checkable property (if Variant() it means the property is unused).

File: scene/2d/audio_stream_player_2d.cpp
Comment (reduz) [reply scene/2d/audio_stream_player_2d.cpp:39]: I think we have to just fuse all 3 in a common internal class at this rate, I am not sure where I would place it.

File: scene/2d/audio_stream_player_2d.cpp
Comment (reduz) [reply scene/2d/audio_stream_player_2d.cpp:470]: Variant is null depends on the parameter, not all stream parameters will be necessarily exposed as checkable.

Comment (reduz) [reply scene/2d/audio_stream_player_2d.cpp:470]: Okay I figured it out, should not save it now.

**Comment by reduz** [comment]
Okay I think this should be ready now.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (KoBeWi) [reply modules/minimp3/audio_stream_mp3.cpp:257]: I mean Parameter constructor is defined like this
`Parameter(const PropertyInfo &p_info = PropertyInfo(), const Variant &p_default_value = Variant())`
The `variant()` you use here is part of the constructor and it's unnecessary.

**Comment by KoBeWi** [comment]
Now the parameter is always null and can't be changed.

File: doc/classes/AudioStream.xml
Comment (adamscott) [inline doc/classes/AudioStream.xml:32]:
```suggestion
			<return type="Dictionary[]" />
```

Based on the failing CI prompt.

Comment (ellenhp) [reply doc/classes/AudioStream.xml:32]: Might make sense to also slightly tweak the description below after this change since "array contains dictionaries" will be redundant.

**Review by ellenhp — APPROVED** [review]
LGTM for audio. I have not built or tested so it would be good if someone else could confirm that they've played with it, but the approach seems sound and I think this'll make a lot of people really happy.

**Comment by reduz** [comment]
@ellenhp Oh, very happy to see you back Ellen and thanks for the review! I still have a few bugs to fix, but should be able to get this one done soon.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (adamscott) [inline modules/minimp3/audio_stream_mp3.cpp:49]: I don't know about this. Maybe the parameter for `looping` shouldn't be a binary, but rather an enum: DEFAULT, ON, OFF

In my mind, we should be able to let the looping parameter to DEFAULT and loop the sample if the stream itself is set to looping.

File: scene/2d/audio_stream_player_2d.cpp
Comment (adamscott) [reply scene/2d/audio_stream_player_2d.cpp:39]: Maybe a .inc file for common classes?

File: scene/2d/audio_stream_player_2d.cpp
Comment (adamscott) [inline scene/2d/audio_stream_player_2d.cpp:480]:
```suggestion
	if (!I) {
		return false;
	}
	
	ParameterData &pd = I->value;
	pd.value = p_value;
	for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
		playback->set_parameter(pd.path, pd.value);
	}
	if (p_value.is_null()) {
		playback_parameters.erase(p_name);
	}
	return true;
```

File: modules/minimp3/audio_stream_mp3.cpp
Comment (KoBeWi) [reply modules/minimp3/audio_stream_mp3.cpp:49]: The property itself is checkable, like theme overrides. So you can choose whether use the default or a custom value.

File: scene/3d/audio_stream_player_3d.cpp
Comment (adamscott) [inline scene/3d/audio_stream_player_3d.cpp:867]: Same return early comment here.

File: scene/audio/audio_stream_player.cpp
Comment (adamscott) [inline scene/audio/audio_stream_player.cpp:138]: Same return early comment here.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (adamscott) [reply modules/minimp3/audio_stream_mp3.cpp:49]: image

Just built the PR to make sure and you're right.

Comment (KoBeWi) [reply modules/minimp3/audio_stream_mp3.cpp:49]: Right, the original screenshot is outdated.

**Comment by reduz** [comment]
> Now the parameter is always null and can't be changed.

Ah shit, I thought It worked, let me test again :laughing:

**Comment by reduz** [comment]
Yeah nevermind, I was misled thinking it worked, can't do it like this. Guess the null will have to stay if unassigned, else the logic gets more hairy due to using checkable logic.

**Comment by KoBeWi** [comment]
Maybe you can handle it using `_validate_property()`. The null only appears after toggling the parameter, so you get a new value that can't be removed anymore.

**Comment by reduz** [comment]
@KoBeWi The problem is that this is very parameter dependant, and this null comes from a checkable parameter in the OGG/MP3 streams, other parameters may not necessarily be like this.
I could still add a hack in _validate_property and make a parameter not serializable if its null, but it feels too workaroundish and not sure if this is worth it.

File: scene/3d/audio_stream_player_3d.cpp
Comment (reduz) [reply scene/3d/audio_stream_player_3d.cpp:867]: I think we will need to have a discussion about these things because I don't think we need to unify so much in style for this type of code. To me early return is not necessarily more readable always. I will do it for this PR, but I think we need to lax the review requirements a bit.

**Comment by KoBeWi** [comment]
We'll get a report as soon as someone finds out a parameter in their scene that they can't remove.

File: scene/2d/audio_stream_player_2d.cpp
Comment (KoBeWi) [reply scene/2d/audio_stream_player_2d.cpp:39]: #87061

**Comment by Mickeon** [comment]
The fact that `null` is saved as is now needs to be addressed before this gets merged, otherwise it's not going to leave a great impression, in my opinion. The feature would feel underbaked right away. There's many cases of properties saving when they don't need to, and I wouldn't want this to contribute to the pile, either.

**Comment by KoBeWi** [comment]
> I could still add a hack in _validate_property and make a parameter not serializable if its null, but it feels too workaroundish and not sure if this is worth it.

Actually this should check if the value is equal to the default value defined in the parameter.

**Review by KoBeWi — APPROVED** [review]
Well, aside from the null issue, looks fine.
I'll try to address it when I update #87061.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (MJacred) [inline modules/minimp3/audio_stream_mp3.cpp:166]: In line 191, add this
```c++
mp3s->looping = loop;
```
to inherit looping setting from stream and be in sync with ogg vorbis.

**Review by MJacred — CHANGES_REQUESTED** [review]
Aside from the suggested change, the audio side lgtm.

File: modules/minimp3/audio_stream_mp3.cpp
Comment (reduz) [reply modules/minimp3/audio_stream_mp3.cpp:166]: ah that line should actually be removed instead, since its ignored.

Comment (MJacred) [reply modules/minimp3/audio_stream_mp3.cpp:166]: Ah, I get it. If the overwrite is deactivated, the inspector can't read `looping` and the audio player mustn't read it.

File: modules/vorbis/audio_stream_ogg_vorbis.cpp
Comment (MJacred) [inline modules/vorbis/audio_stream_ogg_vorbis.cpp:405]:
```suggestion
```

As noted in the previous review: this assignment is ignored anyway.

**Review by MJacred — CHANGES_REQUESTED** [review]
Follow-up from previous review

**Comment by reduz** [comment]
Alright, should be ready to merge once it passes CI.

**Comment by akien-mga** [comment]
Thanks!

**Comment by reduz** [comment]
:tada: 

Thanks all involved for giving me feedback on this!

**Review by Mickeon — COMMENTED** [review]
This feedback has been in a **pending** state for a really, really long time because I forgot to submit it. I [mentioned it on RocketChat a while back](https://chat.godotengine.org/channel/audio?msg=CWBRzKpCZeSiYtvA6) and it was potentially also later discussed, offhandendly, in a meeting on April 7th 2025.


I thought that, while out-of-date, I should submit this review now for the sake of transparency.

**Comment by Nichgon** [comment]
I noticed that `parameters/looping` is only available when stream (AudioStream) is from Ogg and MP3 but not for WAV. It would be great to have this parameter for WAV files as well.

**Pull request merged**
